File: comms/server/networks/trace_pin.py
################################################################################
# Comms Fiber Trace Node
################################################################################
# Copyright: IQGeo Limited 2010-2023

from myworldapp.core.server.base.core.myw_error import MywError
from myworldapp.core.server.networks.myw_trace_node import MywTraceNode


class TracePin:
    """
    A pin within a pin trace node

    Used for modelling paths"""

    def __init__(self, node, pin):
        """
        Init slots of self

        NODE is a PinTraceNode"""

        self.node = node
        self.pin = pin

    def __ident__(self):
        """
        String used to identify self in test results etc
        """

        return "{}({},{})".format(self.__class__.__name__, self.node.node_id, self.pin)

    def definition(self, full=False):
        """
        JSON-serialisable form of self
        """

        feature = self.node.feature

        if self.node.type == "segment":
            type = "cable"
            feature = feature._field("cable").rec()
            desc = "{}#{}".format(feature.name, self.pin)
        else:
            type = "port"
            desc = "{}#{}:{}".format(feature.name, self.node.pins.side.upper(), self.pin)

        defn = {
            "type": type,
            "id": self.text(),
            "feature": self.node.feature._urn(),
            "title": feature._title(),
            "desc": desc,
            "side": self.node.pins.side,
            "pin": self.pin,
        }

        if full:
            defn["coords"] = self.node.coordsFromRoot()

        return defn

    def text(self):
        """
        String representation of self for trace tree
        """

        if self.node.type == "segment":  # ENH: Move to node subclass
            name = self.node.feature._field("cable").rec().name
        else:
            name = self.node.feature.name

        return "{}#{}:{}".format(name, self.node.pins.side, self.pin)

File: comms/native/services/networks/TracePin.js
//###############################################################################
// Comms Fiber Trace Node
//###############################################################################
// Copyright: IQGeo Limited 2010-2023

//import OrderedDict from 'collections'
//import {MywError} from 'myWorld-native-services'
//import {MywTraceNode} from 'myWorld-native-services'

import { MywClass } from 'myWorld-base';

/**
 * A pin within a pin trace node
 *
 * Used for modelling paths
 */
class TracePin extends MywClass {
    /**
     * Init slots of self
     *
     * NODE is a PinTraceNode
     */
    constructor(node, pin) {
        super();
        this.node = node;
        this.pin = pin;
    }

    /**
     * JSON-serialisable form of self
     */
    async definition(full = false) {
        let feature = this.node.feature;
        let type = '';
        let desc = '';

        if (this.node.type == 'segment') {
            type = 'cable';
            feature = await feature.followRef('cable');
            desc = `${feature.properties.name}#${this.pin}`;
        } else {
            type = 'port';
            desc = `${feature.properties.name}#${this.node.pins.side.toUpperCase()}:${this.pin}`;
        }

        const defn = {
            type: type,
            id: await this.text(),
            feature: this.node.feature.getUrn(),
            title: feature.myw.title,
            desc: desc,
            side: this.node.pins.side,
            pin: this.pin
        };

        if (full) {
            defn['coords'] = this.node.coordsFromRoot();
        }

        return defn;
    }

    /**
     * String representation of self for trace tree
     */
    async text() {
        let name = '';

        if (this.node.type == 'segment') {
            const feature = await this.node.feature.followRef('cable');
            name = feature.properties.name;
        } else {
            name = this.node.feature.properties.name;
        }

        return `${name}#${this.node.pins.side}:${this.pin}`;
    }
}

export default TracePin;
